### ISSUE TITLE
Possibility to optimize global app-specific and frame-specific state binding?

### ISSUE THREAD

**Issue by juj** [gpuweb/gpuweb#1414]
Renderers have state that they bind to GPU at different granularities: global app/scene specific  data, per-frame data, per-render target data, per-camera data, per-light data, per-object/draw call data.

In the current spec, when one invokes
```
const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
```
that will start from a "clean state", i.e. reset all previously bound state with calls to earlier `passEncoder.set{Pipeline,IndexBuffer,VertexBuffer,Viewport,ScissorRect,BlendColor,StencilReference,BindGroup` functions. As result, if one either has global application wide state, or per-frame state that should persist across render passes, they need to be rebound at each pass.

Not 100% familiar with the native APIs on this front, do they also require resetting all state in the beginning of each render pass?

It would be more convenient to be able to once at page init set bind groups that represent global app/scene specific state, and also in the beginning of each frame to set the frame-specific state, rather than need to rebind such state after each new pass start.

I wonder if that is something that is lacking compared to native parity? Would it make sense to consider this kind of optimization?

**Comment by kvark** [comment]
Render passes are expected to be one of the heaviest things you can do every frame. They have both CPU and GPU cost. You aren't expected to have *many* render passes, and therefore, the possible overhead of re-binding the global state at each render pass is negligible. I.e. if you have 100+ render passes, the state binding will be the least of your problems, I doubt you'll notice it in the profiles, dominated by other render pass costs.

TL;DR: I don't think this is a problem that needs solving right now

**Comment by kainino0x** [comment]
Note, since bind groups just hold references to their resources, you can squish "global" data into one of your other bind groups (e.g. the one that has the camera) and pay zero cost because no extra setBindGroup call is needed. (Updating the data simply requires updating the contents of a buffer/texture, no new bind group.)

For the rest of them, I estimate the chances you actually need them across multiple render passes as:
- setIndexBuffer/VertexBuffer/BlendColor: ~0
- setStencilReference: not really sure, but ~0?
- setPipeline: rare
- setViewport: sometimes, but low cost

**Comment by juj** [comment]
Good points.

I was not thinking at this from the perspective of "optimize performance", but "optimize for clarity/code structure", where it would be more straightforward to init these once at page load/scene load/frame start, and not have to have functions that rebind needed base state again at each pass start.

Although I have not written much WebGPU code to observe how tedious that would be. Also not sure what the implications would be to the implementing backends, hence the question.

**Comment by Kangz** [comment]
I think this conversation is resolved?

**Comment by juj** [comment]
Yeah, indeed, closing.
